openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -408,34 +417,66 @@ def _verify_user(request, response, userid_in_session):
+
             if hasattr(request.user, 'real_user'):
                 # If a view overrode the request.user with a masqueraded user, this will
-                #   revert/clean-up that change during response processing.
+                # revert/clean-up that change during response processing.
                 request.user = request.user.real_user
-            # The user at response time is expected to be None when the user
-            # is logging out.  We won't log that.
-            if request.safe_cookie_verified_user_id != request.user.id and request.user.id is not None:
-                log.warning(
-                    (
-                        "SafeCookieData user at request '{}' does not match user at response: '{}' "
-                        "for request path '{}'"
-                    ).format(  # pylint: disable=logging-format-interpolation
-                        request.safe_cookie_verified_user_id, request.user.id, request.path,
-                    ),
-                )
-                set_custom_attribute("safe_sessions.user_mismatch", "request-response-mismatch")
-            # The user session at response time is expected to be None when the user
-            # is logging out.  We won't log that.
-            if request.safe_cookie_verified_user_id != userid_in_session and userid_in_session is not None:
-                log.warning(
-                    (
-                        "SafeCookieData user at request '{}' does not match user in session: '{}' "
-                        "for request path '{}'"
-                    ).format(  # pylint: disable=logging-format-interpolation
-                        request.safe_cookie_verified_user_id, userid_in_session, request.path,
-                    ),
-                )
-                set_custom_attribute("safe_sessions.user_mismatch", "request-session-mismatch")
+
+            session_at_response = 'None'
+            if hasattr(request, 'session') and hasattr(request.session, 'session_key'):
+                session_at_response = request.session.session_key
+
+            if is_request_response_mismatch or is_request_session_mismatch:
+                if not cache.touch(USER_MISMATCH_CACHE_KEY, 300):
+                    cache.set(USER_MISMATCH_CACHE_KEY, True, 300)
+                session_has_changed = not request.safe_cookie_session_id == session_at_response
+                # remove the stored old session id for security
+                del request.safe_cookie_session_id
+
+                # for security, don't log the actual session id, just whether or not it changed
+                log_session_suffix = 'Session id has changed.' if session_has_changed else 'Session id has not changed'
+
+                if not is_request_session_mismatch:
+                    log.warning(
+                        (
+                            "SafeCookieData user at request '{}' does not match user in response: '{}' "
+                            "for request path '{}'. {}"
```

Comment:
Nit: It's not really in the response, it's `request.user.id` at response _time_ -- the original message was a bit telegraphic here.